Mathlib/Algebra/SkewMonoidAlgebra/Basic.lean

```diff
@@ -423,8 +444,394 @@ theorem sum_congr {f : SkewMonoidAlgebra k G} {M : Type*} [AddCommMonoid M] {g
     (h : ∀ x ∈ f.support, g₁ x (f.coeff x) = g₂ x (f.coeff x)) :
     f.sum g₁ = f.sum g₂ := Finset.sum_congr rfl h
 
+theorem induction_on {p : SkewMonoidAlgebra k G → Prop} (f : SkewMonoidAlgebra k G)
+    (zero : p 0) (single : ∀ g a, p (single g a)) (add : ∀ f g :
+    SkewMonoidAlgebra k G, p f → p g → p (f + g)) : p f := by
+  rw [← sum_single f, sum_def']
+  exact Finset.sum_induction _ _ add zero (by aesop)
+
+/--
+Slightly less general but more convenient version of `SkewMonoidAlgebra.induction_on`. -/
+theorem induction_on' [instNonempty : Nonempty G] {p : SkewMonoidAlgebra k G → Prop}
+    (f : SkewMonoidAlgebra k G) (single : ∀ g a, p (single g a)) (add : ∀ f g :
+    SkewMonoidAlgebra k G, p f → p g → p (f + g)) : p f :=
+  induction_on _ (by simpa using single (Classical.choice instNonempty) 0) single add
+
+/-- If two additive homomorphisms from `SkewMonoidAlgebra k G ` are equal on each `single a b`,
+then they are equal. -/
+@[ext high]
+theorem addHom_ext {M : Type*} [AddZeroClass M] {f g : SkewMonoidAlgebra k G →+ M}
+    (h : ∀ a b, f (single a b) = g (single a b)) : f = g := by
+  ext p; induction p using SkewMonoidAlgebra.induction_on <;> simp_all
+
 end sum
 
+section mapDomain
+
+/-- Given `f : G → G'` and `v : SkewMonoidAlgebra k G`, `mapDomain f v : SkewMonoidAlgebra k G'`
+is the finitely supported function whose value at `a : G'` is the sum of `v x` over all `x`
+such that `f x = a`. -/
+def mapDomain {G' : Type*} (f : G → G') (v : SkewMonoidAlgebra k G) : SkewMonoidAlgebra k G' :=
+  v.sum fun a ↦ single (f a)
+
+@[simp]
+lemma toFinsupp_mapDomain {G' : Type*} (f : G → G') (v : SkewMonoidAlgebra k G) :
+    (mapDomain f v).toFinsupp = Finsupp.mapDomain f v.toFinsupp := by
+  simp_rw [mapDomain, Finsupp.mapDomain, toFinsupp_sum', single]
+
+@[simp]
+theorem mapDomain_id {v : SkewMonoidAlgebra k G} : mapDomain id v = v := sum_single _
+
+theorem mapDomain_comp {G' G'' : Type*} {f : G → G'} {g : G' → G''} {v : SkewMonoidAlgebra k G} :
+    mapDomain (g ∘ f) v = mapDomain g (mapDomain f v) :=
+  ((sum_sum_index (single_zero <| g ·) (single_add <| g ·)).trans
+    (sum_congr fun _ _ ↦ sum_single_index (single_zero _))).symm
+
+theorem mapDomain_zero {G' : Type*} {f : G → G'} :
+    mapDomain f (0 : SkewMonoidAlgebra k G) = (0 : SkewMonoidAlgebra k G') :=
+  sum_zero_index
+
+theorem mapDomain_add {G' : Type*} {f : G → G'} {v₁ v₂ : SkewMonoidAlgebra k G} :
```

Comment:
What about bundling `mapDomain` as an add hom? In this way `mapDomain_sum` (and maybe something else) can be removed.

Reply:
Thanks for the review!
I have bundled it as an add hom, keeping the name `mapDomain`. I think this is a good change, but I should point out that for `MonoidAlgebra`, mathlib currently has [MonoidAlgebra.mapDomain](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/MonoidAlgebra/MapDomain.html#MonoidAlgebra.mapDomain), which is just a function, and then [MonoidAlgebra.mapDomainRingHom](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/MonoidAlgebra/MapDomain.html#MonoidAlgebra.mapDomainRingHom); we plan to add the `RingHom` version for `SkewMonoidAlgebra` in a future PR.